### DISCUSSION TITLE
tRPC Studio/Playground?

### DISCUSSION THREAD

**Discussion by albingroen** [trpc/trpc#1372] [Q&A]
Wouldn't it be pretty cool to have something similar to Apollo playground, but for tRPC?

I would be happy to work on this if there's any interest from others 😊

**Comment by sachinraja** [comment]
I've actually been working on this over the holidays. I wanted to keep it under wraps until I get a few more features in, but here's a preview of it for those interested:
https://user-images.githubusercontent.com/58836760/148662415-91310d03-73b1-49dd-a67d-70640c1906f6.mp4. The editor is backed by [CodeMirror](https://github.com/codemirror/CodeMirror) and the typescript language server (all in your browser!).

Edit: [Now released!](https://github.com/sachinraja/trpc-playground)

**Reply by albingroen** [reply]
Cool! Unfortunately I can't seem to get the video to load though?

**Reply by sachinraja** [reply]
It doesn't look like GitHub created an embed for it. You should be able to click on the link and watch it though. I shared it on the Discord too and it seemed to work just fine (it even created an embed for the video).

**Reply by albingroen** [reply]
Ah okay, it seemed to be a Safari issue...

**Reply by sachinraja** [reply]
It's almost ready for use, but there's still a bit of tweaking I want to do before releasing it. Let me know if you're interested in collaborating though and I'll invite to the repo!

**Reply by albingroen** [reply]
Yeah, sure - feel free to send me an invite!

**Reply by sachinraja** [reply]
Released at https://github.com/sachinraja/trpc-playground

**Reply by KATT** [reply]
🤩🤩
Please add it to the awesome list section on the website!

**Reply by sachinraja** [reply]
PR at #1447!

**Comment by albingroen** [comment]
Here are a couple of mockup screens for what I have in mind ☁️

![MacBook Pro 14_ - 1](https://user-images.githubusercontent.com/19674362/148663066-0ac6642b-9d45-4864-9c9d-8bcb4e4c321f.png)
![MacBook Pro 14_ - 2](https://user-images.githubusercontent.com/19674362/148663067-fde2b97e-5100-4468-a04a-a36c46205bdf.png)
![MacBook Pro 14_ - 3](https://user-images.githubusercontent.com/19674362/148663068-868d6720-54eb-44e6-be9c-0e874c5ee5cf.png)

**Reply by sachinraja** [reply]
Woah I really like the design of this. It looks similar to GraphQL Playground. I chose to go a different way in my playground that I think fits tRPC better, which takes more inspiration from [the Prisma playground](https://qc.prisma-adp.vercel.app/) (the main CodeMirror extensions are also from there).

**Reply by albingroen** [reply]
Yeah, I noticed you took a bit of a different approach. I do think that's also very cool though!

**Reply by Explosion-Scratch** [reply]
I like this version a bit better than sachinraja's version tbh

**Comment by mmkal** [comment]
We also built a mini-ui for invoking procedures from a web app. More aimed at ops folks than developers - it serves the schemas of the (whitelisted) queries and mutations as json-schema (using zod-to-json-schema), then on the frontend renders forms for each procedure using react-json-schema-form. So running the procedures is point and click, no need for ops people to learn JavaScript or even JSON. If there'd be interest we could share it too

**Reply by albingroen** [reply]
Yeah, I think I like this approach better, to be honest. It'd be interesting to hear if @KATT has any thoughts or plans for a official one, instead of everyone building their own versions?

**Reply by sachinraja** [reply]
> no need for ops people to learn JavaScript or even JSON

I think it's safe to assume that people using tRPC or tRPC apis know basic JavaScript. I think it makes sense to keep writing JS in the playground. Why should it change? With TS you can provide the same fully typed experience in the browser and code like you normally do. 

It might make sense to combine these approaches though and perhaps be able to quickly generate a query. I think it would be good to provide a good experience for both developing a tRPC api and exposing a playground for users to make queries against your API.

**Reply by mmkal** [reply]
>I think it's safe to assume that people using tRPC or tRPC apis know basic JavaScript

Not in our case. Our ops team are using web pages with input boxes, drop-downs, checkboxes etc. and then clicking submit. They don't even know what trpc is. But to be honest we're talking about slightly different use cases. A developer playground vs autogenerated admin UI. Both have plenty of value. 

Maybe there could be a section of the repo for blessed community add-ons, some of which might be explicitly marked as experimental. There are almost certainly edge cases our admin tool doesn't cover yet, it's not library-quality right now, but it'd still be great to work on it collaboratively.

**Reply by sachinraja** [reply]
An `awesome-trpc` section would be great. Would be nice to feature ecosystem utilities and open up collaboration.

**Reply by KATT** [reply]
> An `awesome-trpc` section would be great. Would be nice to feature ecosystem utilities and open up collaboration.

Your wish is my command -> https://github.com/trpc/trpc/pull/1379

**Comment by umar-ahmed** [comment]
If there was a built-in playground, it would be nice for those using the next.js adapter to be able to change the API route from `/api/trpc/[trpc].js` to an optional catch-all, ie. `/api/trpc/[[...trpc]].js` and have that automatically open the playground, similar to how Apollo and other graphql servers do.

**Reply by iway1** [reply]
this is a great idea, I'm working on https://trpcpanel.io which is still pretty early in development - hoping I can have something like this where the dev has the option to start the UI without changing any server code but right now it has to inspect the router JS object so IDK how feasible it is

**Comment by tanlethanh** [comment]
Hi everyone, I've built a playground UI [https://trpc-studio.vercel.app/](https://trpc-studio.vercel.app/). 

The API needs to be wrapped by `@trpc-playground/introspection` to expose an introspection endpoint just like GraphQL. Then it's ready to be load by the UI with an auto suggestion JSON editor and a cool schema UI editor.

You can checkout the repo [https://github.com/tanlethanh/trpc-studio](https://github.com/tanlethanh/trpc-studio). It's just a simple version with basic functionalities, PRs are welcomed. Hope you guys find it useful!
